### PR TITLE
Client authentication filter

### PR DESCRIPTION
**Pull request by tomvandenberge**
This adds a rather simple filter for client authentication (#5) for the Token Endpoint. It can currently only be used for the Client Credentials Grant.

This PR contains some files that shouldn't be there. This is probably a result of my earlier failed attempts to squash/rebase for another PR. I hope it doesn't mess it up too badly. This PR should only contain the filter, the tests for the filter, and three exception classes.

### CODE REVIEW

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        return credentials;
+    }
+
+    private String[] extractBasicAuthenticationCredentials(HttpServletRequest request) {
+        String header = request.getHeader("Authorization");
+        if (header != null && header.toLowerCase().startsWith("basic ")) {
+            return extractAndDecodeHeader(header, request);
+        }
+        return null;
```
Comment (Tandolf) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:73]: returning null is a risk and makes for the need to do a lot of nasty null checks later in the code, try leveraging the return of `Optional<T>` instead.

Comment (Captain-P-Goldfish) [reply src/main/java/sample/ClientCredentialsAuthenticationFilter.java:73]: either an optional or throw directly an UnauthorizedException
Alternatively extract the whole header and build than a method that choses the authentication implementation by the auth-scheme in the Authorization-header.

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        HttpServletRequest req = (HttpServletRequest) request;
+        if (req.getRequestURI().equals("/oauth/token")) {
+            String grantType = req.getParameter("grant_type");
+            if ("client_credentials".equals(grantType)) {
+                String[] credentials = extractCredentials(req);
+                String clientId = credentials[0];
+                String clientSecret = credentials[1];
+                if (!clientId.equals("myclientid") || !clientSecret.equals("myclientsecret")) {
+                    throw new BadCredentialsException("Invalid client credentials");
+                }
```
Comment (Captain-P-Goldfish) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:49]: still pretty much hardcoded

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        if (clientId != null) {
+            String clientSecret = request.getParameter("client_secret");
+            if (clientSecret == null) {
+                throw new InvalidRequestException("No client_secret parameter provided");
+            }
+            return new String[] { clientId, clientSecret };
+        }
+        return null;
+    }
+
```
Comment (Captain-P-Goldfish) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:87]: the pitfals of mockito.
you will surely never get a request in which you will be able to get the client_secret as a parameter from the request-object. The client_secret is always encoded in the Authorization header.
Besides you wrote that you only implemented the *client credentials grant*. 
https://tools.ietf.org/html/rfc6749#section-4.4.2
RFC6749 does neither define the client_id nor the client_secret in the "*client credentials grant*" "*access token request*".
The only required parameter is "*grant_type*"

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    // Taken from BasicAuthenticationFilter (spring-security-web)
+    private String[] extractAndDecodeHeader(String header, HttpServletRequest request) {
```
Comment (Captain-P-Goldfish) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:89]: why give the header name as parameter? At least I do only know one header ("Authorization") that may contain the content that is necessary for this method to operate successfully. If I am wrong please correct me.

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        try {
+            decoded = Base64.getDecoder().decode(base64Token);
+        }
+        catch (IllegalArgumentException e) {
+            throw new BadCredentialsException("Failed to decode basic authentication token");
+        }
+
+        String token = new String(decoded, getCredentialsCharset(request));
+
+        int delim = token.indexOf(":");
```
Comment (Captain-P-Goldfish) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:102]: why not "*split*" and spare yourself the substring methods below?

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+        if (delim == -1) {
+            throw new BadCredentialsException("Invalid basic authentication token");
+        }
+        return new String[] { token.substring(0, delim), token.substring(delim + 1) };
+    }
```
Comment (Captain-P-Goldfish) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:108]: You got the same problem here as I mentioned in this pull request:
https://github.com/spring-projects-experimental/spring-authorization-server/pull/63

File: build.gradle [hunk 7]
```diff
@@ -0,0 +1,25 @@
+plugins {
```
Comment (jgrandja) [inline build.gradle:1]: Please move the code into the [minimal](https://github.com/spring-projects-experimental/spring-authorization-server/blob/master/samples/boot/minimal/spring-authorization-server-samples-boot-minimal.gradle) sample and remove this gradle build.

File: gradle/wrapper/gradle-wrapper.properties [hunk 8]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
```
Comment (jgrandja) [inline gradle/wrapper/gradle-wrapper.properties:1]: Please remove as this has been added in #10

File: gradlew [hunk 9]
```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env sh
```
Comment (jgrandja) [inline gradlew:1]: Please remove as this has been added in #10

File: gradlew.bat [hunk 10]
```diff
@@ -0,0 +1,103 @@
+@rem
```
Comment (jgrandja) [inline gradlew.bat:1]: Please remove as this has been added in #10

File: settings.gradle [hunk 11]
```diff
@@ -0,0 +1 @@
+rootProject.name = 'spring-authorization-server'
```
Comment (jgrandja) [inline settings.gradle:1]: Please remove as this has been added in #10

File: src/main/java/sample/Application.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package sample;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class Application {
```
Comment (jgrandja) [inline src/main/java/sample/Application.java:23]: Please remove as this exists in the [minimal](https://github.com/spring-projects-experimental/spring-authorization-server/blob/master/samples/boot/minimal/spring-authorization-server-samples-boot-minimal.gradle) sample.

File: src/main/java/sample/BadCredentialsException.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sample;
+
+public class BadCredentialsException extends RuntimeException {
```
Comment (jgrandja) [inline src/main/java/sample/BadCredentialsException.java:19]: Let's add the minimal amount of code to get things working for this initial iteration. Please remove this and instead the `Filter` can set the status if it catches an exception from the `AuthenticationManager`.

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    private Charset credentialsCharset = UTF_8;
+
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+
+        HttpServletRequest req = (HttpServletRequest) request;
+        if (req.getRequestURI().equals("/oauth/token")) {
+            String grantType = req.getParameter("grant_type");
+            if ("client_credentials".equals(grantType)) {
```
Comment (jgrandja) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:43]: Please remove the `grant_type` check. We should authenticate if the target is the token endpoint and there are credentials in the request.

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+                String[] credentials = extractCredentials(req);
+                String clientId = credentials[0];
+                String clientSecret = credentials[1];
+                if (!clientId.equals("myclientid") || !clientSecret.equals("myclientsecret")) {
+                    throw new BadCredentialsException("Invalid client credentials");
+                }
+                req.getSession().setAttribute("authentication", true);
```
Comment (jgrandja) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:50]: Please remove this as the client `Authentication` should be set in the `SecurityContext` if authentication passes.

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+            }
+        }
+
+        chain.doFilter(request, response);
+    }
+
+    private String[] extractCredentials(HttpServletRequest request) {
+        String[] credentials = extractBasicAuthenticationCredentials(request);
+        if (credentials == null) {
+            credentials = extractParameterAuthenticationCredentials(request);
```
Comment (jgrandja) [inline src/main/java/sample/ClientCredentialsAuthenticationFilter.java:60]: For this initial iteration, we will support HTTP Basic only. All other authentication schemes should be removed.

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java (hunk 3 above)
Comment (jgrandja) [reply src/main/java/sample/ClientCredentialsAuthenticationFilter.java:87]: @Captain-P-Goldfish Please see [4.4.2.  Access Token Request]( https://tools.ietf.org/html/rfc6749#section-4.4.2)

> The client MUST authenticate with the authorization server as described in Section 3.2.1.

File: src/main/java/sample/InvalidRequestException.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sample;
+
+public class InvalidRequestException extends RuntimeException {
```
Comment (jgrandja) [inline src/main/java/sample/InvalidRequestException.java:19]: Let's add the minimal amount of code to get things working for this initial iteration. Please remove this and instead the `Filter` can set the status if it catches an exception from the `AuthenticationManager`.

File: src/test/java/sample/ApplicationTest.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+package sample;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.context.ApplicationContext;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest
+public class ApplicationTest {
```
Comment (jgrandja) [inline src/test/java/sample/ApplicationTest.java:26]: Please remove as this exists in the [minimal](https://github.com/spring-projects-experimental/spring-authorization-server/blob/master/samples/boot/minimal/spring-authorization-server-samples-boot-minimal.gradle) sample.

File: src/main/java/sample/ClientCredentialsAuthenticationFilter.java (hunk 3 above)
Comment (Captain-P-Goldfish) [reply src/main/java/sample/ClientCredentialsAuthenticationFilter.java:87]: right but this does not nullify my issue does it?
The client_id and client_secret must be extracted from the Authorization-header not from the requests parameters

**Comment by tomvandenberge** [comment]
@jgrandja I'm struggling with how to correctly respond to your review comments. I'm not very familiar with working with pull requests and reviews in github. I think I've resolved the things you've mentioned, and committed and pushed them. But I'm really not sure if you're able to see them now. I have been looking for some more details on the reviewing process, but couldn't find anything useful. Please let met know.

**Comment by jgrandja** [comment]
@tomvandenberge I'm still seeing some files that should have been removed as per my last review comments.

Can you be more specific on the struggles you are having?

Have you reviewed [contributing](https://github.com/spring-projects-experimental/spring-authorization-server#contributing), as it provides links to working with pull requests, squashing commits using interactive rebasing, etc.

Please let me know exactly what you need help with and I'll point you in the right direction.

**Comment by tomvandenberge** [comment]
@jgrandja I don't understand why you want me to remove e.g. build.gradle, and the other gradle files. Without these, gradle doesn't work, or am I missing something?

**Comment by tomvandenberge** [comment]
@jgrandja Thanks for pointing me to these docs. I'm having a better understanding of rebasing and squashing now, and I'll try it with my next changes. My previous attempt to squash commits in Intellij was a disaster ;-) 

Is it recommended to rebase (instead of merge) when pulling updates from the upstream repository?

**Comment by jgrandja** [comment]
@tomvandenberge 

> I don't understand why you want me to remove e.g. build.gradle, and the other gradle files

The project has been setup with gradle multi modules via 5ed7c8f. And the sample project already has a [gradle file](https://github.com/spring-projects-experimental/spring-authorization-server/tree/master/samples/boot/minimal)

> Is it recommended to rebase (instead of merge)

Yes, please rebase on top of master

**Comment by tomvandenberge** [comment]
@jgrandja it must be my lack of experience with gradle. When I remove `gradlew`, `gradle/wrapper/gradle-wrapper.properties` and `settings.gradle` as per your comments, I can't run gradle anymore.
The [build instructions](https://github.com/spring-projects-experimental/spring-authorization-server#building-from-source) also explicitly state that there should be a `gradlew` in the root of the source.

**Comment by Tandolf** [comment]
@tomvandenberge what they mean is that you should remove the file from the commit, not from your harddrive itself. You still need them localy to run your build, but they should not be commited in your PR.

**Comment by tomvandenberge** [comment]
@Tandolf Thanks for your clarification. I'm afraid my next question is how do I remove specific files from a commit?

**Comment by Tandolf** [comment]
@tomvandenberge you use the command [git rm](https://git-scm.com/docs/git-rm). And i believe for you to understand how you ended up in this situation. Before you commit you should **always** check with the `git status` command what git wants to commit. If git wants to commit files you havn't edited, you should look into why this is so.

You should **never** commit files you havn't touched. My belive is that you are facing the "CRLF" issue. If you have no idea what im talking about i recommend you reading up on what "CR" and "LF" is, the problems with it and how git can handle it, because it is too long to explain here.

**Comment by tomvandenberge** [comment]
@Tandolf These files were not added or modified accidentally. I've initially added them for another PR, and while I was working on this one, somebody changed them. I merged  these upstream changes into my branch, which I maybe shouldn't have done (?), or maybe I should have used rebasing instead of merging when I did this, I don't know if that would have made a difference.

I've tried to remove the files with `git rm` as you said (isn't this exactly the same as removing the files locally, and then committing that?). Trying to push this to my origin branch didn't work. Apparently I had to pull first. After a lot of hassling, the files were no longer in my repository. Then the builds in the upstream Spring repo failed, because the gradle files were no longer there (which makes sense in my view, because `git rm` removed them). To fix this, I added them again, which brings me back to square one again.

No doubt that the git workflow I'm using is causing these problems. That's what probably got me into this mess. Any help to get me out of it is greatly appreciated!

**Comment by jgrandja** [comment]
@tomvandenberge 

> When I remove gradlew, gradle/wrapper/gradle-wrapper.properties and settings.gradle as per your comments, I can't run gradle anymore.

The gradle script already exists in the [project root](https://github.com/spring-projects-experimental/spring-authorization-server). You can run the build (from project root) using `./gradlew build`, which will build ALL the modules. Or you can run a specific module, eg. `./gradlew :spring-authorization-server-samples-boot-minimal:build`.

I'm seeing 20 commits in this PR, which most of them are in master. I think the best approach to get back on track with this PR, is to start from `master` and then manually add your changes into 1 commit. Then you can `--force` push the branch to bring this PR back to 1 commit on top of `master`. Any subsequent changes should be rebased (not merged) on top of master.

**Comment by tomvandenberge** [comment]
@jgrandja 
> The gradle script already exists in the project root. 

I'm aware of that (assuming you mean `gradlew`). My PR also has this file in this location (in [this comment](https://github.com/spring-projects-experimental/spring-authorization-server/pull/64#issuecomment-623149664) I try to explain how it ended up there). In your comments you ask me to remove it. It's still not clear to me what exactly how you want me to this (or why). The same applies to the other gradle-related files.

> I think the best approach to get back on track with this PR, is to start from master and then manually add your changes into 1 commit. 

I'm afraid I don't understand this. What do you mean exactly by "to start from master"? My changes are already in my master branch, and therefore in the PR, so I can't add them again. Or is there some way to "reset" the PR?

Thanks for you patience ;-)

**Comment by jgrandja** [comment]
@tomvandenberge 

> What do you mean exactly by "to start from master"?

I mean start from `origin master` - the master branch in the repository.

> I'm working from my master branch, and not a feature branch

It's recommended to work from a feature branch.

**Comment by tomvandenberge** [comment]
> I mean start from `origin master` - the master branch in this repository.

What should I do in this branch? Could you please be more specific? As I said, it already contains my changes.

**Comment by jgrandja** [comment]
@tomvandenberge I'm suggesting that you create a feature branch and submit that branch for the PR. You could close the existing PR after you submit the new one. I just feel it's easier with this approach as your `master` branch is way off with `origin master`. You will need to `git reset --hard` your `master` to `origin master` after you submit the new PR (with new branch) to sync up your `master`.

Create a new feature branch:

`git checkout -b gh-5-client-auth origin/master`

** Assuming `origin` is the name of the remote in your git setup.

Now, manually apply the changes from your `master` to `gh-5-client-auth` branch. Please ensure you don't copy any of the gradle related files. The only changes should be related to the `Filter`.

After you manually apply the changes then commit.

At this point you will have 1 commit that sits on top of latest commit from `origin/master`.

Now submit a new PR based on this branch. Close the other PR.

All changes going forward should be rebased on (current) `origin/master`.

**Comment by dfcoffin** [comment]
@tomvandenberge @jgrandja 

> I mean start from origin master - the master branch in this repository.

I suspect Joe is asking you to start over after reading from the master branch.  An approach you might consider doing the following steps:

```
git remote add upstream https://github.com/spring-projects-experimental/spring-authorization-server
git checkout repair
git rebase upstream master
```
Then update the new branch with your code changes.  Once the changes use:

`git add <file_name>  Note: wildcard "*" values can be used
git commit -m "msg"`

That way you will retain the files being asked to remove from the submission in your local repository as "untracked" files

Once all of the above is done issue a PR:

`git pull upstream master`

The above will get conflicting issues corrected.

**Comment by dfcoffin** [comment]
@tomvandenberge 

If it helps we can conference and pair to help resolve the issue

**Comment by tomvandenberge** [comment]
@dfcoffin 

> If it helps we can conference and pair to help resolve the issue

That would be very helpful! Can I send you a message on donald.coffin@reminetworks.com to get this started?

**Comment by tomvandenberge** [comment]
@jgrandja 
> I'm suggesting that you create a feature branch and submit that branch for the PR. You could
> close the existing PR after you submit the new one. I just feel it's easier with this approach as 
> your `master` branch is way off with `origin master`

A new branch sounds like the most practical way forward, I agree. What I don't understand is why you think that my `master` branch is off with `origin master`. They are identical. `origin/master` is the remote master branch of my fork. Did you mean `upstream/master`?

Because my `master` and `origin/master` are in sync, creating a new feature branch as you suggest will result in a branch containing all the mess that's currently in my `origin/master`. Therefore I think that the approach you suggest will not work.

**Comment by jgrandja** [comment]
@tomvandenberge 

> What I don't understand is why you think that my master branch is off with origin master. They are identical....Did you mean upstream/master?

Yes, I mean `upstream/master`. Your [master](https://github.com/tomvandenberge/spring-authorization-server/commits/master) branch has 40 commits, whereas [upstream/master](https://github.com/spring-projects-experimental/spring-authorization-server/commits/master) has 29 commits. Your `master` branch has a few "merge" commits, which is causing part of this issue.

**Comment by dfcoffin** [comment]
@tomvandenberge 

> That would be very helpful! Can I send you a message on donald.coffin@reminetworks.com to get this started?

Yes.  I'm in Georgia to help set up a time we are both available

**Pull request closed without merging**